src/Plugin/EntityBrowser/Widget/Webdam.php

```diff
@@ -172,68 +129,247 @@ public function getForm(array &$original_form, FormStateInterface $form_state, a
+      //Store the list of items/assets for rendering later
+      $folder_items = $folder_assets->items;
+    }else{
+      //The webdam root folder is fetched differently because it can only contain subfolders (not assets)
+      $folders = $this->webdam->getTopLevelFolders();
+    }
+    //Add the breadcrumb to the form
+    $form += $this->getBreadcrumb($current_folder, $breadcrumbs);
+    //Add container for assets (and folder buttons)
+    $form['asset-container'] = [
+      '#type' => 'container',
+      //Store the current folder id in the form so it can be retrieved from the form state
+      '#webdam_folder_id' => $current_folder->id,
+    ];
     // Add folder buttons to form
     foreach ($folders as $folder){
       $form['asset-container'][$folder->id] = [
         '#type' => 'button',
         '#value' => $folder->name,
         '#name' => 'webdam_folder',
         '#webdam_folder_id' => $folder->id,
-        '#webdam_parent_folder_id' => $parent,
+        '#webdam_parent_folder_id' => $current_folder->parent,
         '#attributes' => [
           'class' => ['webdam-browser-asset'],
         ],
       ];
     }
     //Assets are rendered as #options for a checkboxes element.  Start with an empty array.
     $assets = [];
-
     //Add to the assets array
     if (isset($folder_items)) {
       foreach ($folder_items as $folder_item) {
         $assets[$folder_item->id] = $this->layoutMediaEntity($folder_item);
       }
     }
-
     // Add assets to form.
     $form['asset-container']['assets'] = [
       '#type' => 'checkboxes',
       '#title' => $this->t('Choose one or more assets'),
       '#title_display' => 'invisible',
       '#options' => $assets,
-      // Multiple assets will only be accepted if the source field allows more than one value.
-      '#multiple' => $field_cardinality != 1 && $this->configuration['multiple'],
       '#attached' => [
         'library' => [
           'media_webdam/webdam',
         ]
       ]
     ];
+    //If the number of assets in the current folder is greater than the number of assets to show per page
+    if($current_folder->numassets > $num_per_page) {
+      //Add the pager to the form
+      $form += $this->getPager($current_folder, $page, $num_per_page);
+    }
     return $form;
   }
 
   /**
    * {@inheritdoc}
    */
   protected function prepareEntities(array $form, FormStateInterface $form_state) {
-    foreach ($form_state->getValue(['upload'], []) as $aid) {
-      if ($aid !== 0) {
-        //TODO: validate form selection
+    $entities = [];
+    $media_bundle = MediaBundle::load($this->configuration['bundle']);
+    $asset_ids = $form_state->getValue(['assets'], []);
+    $assets = $this->webdam->getAssetMultiple($asset_ids);
+    foreach ($assets as $asset) {
+      $entity_values = [
+        'bundle' => $this->configuration['bundle'],
+        'uid' => \Drupal::currentUser()->id(),
+        'langcode' => \Drupal::languageManager()->getCurrentLanguage()->getId(),
```

Comment:
In general, `\Drupal` should only be used in non-OO code (i.e. in a .module file or something). In this case, the current_user service and the language_manager service can be injected via create() and saved to a protected variable. This way, in unit tests, we can pass in a mocked version of these services for testing purposes. Miles can provide more information on how to do this - he's doing a lot of this on the NOAA project.